### DISCUSSION TITLE
bpftrace to monitor cpu and memory usage per process

### DISCUSSION THREAD

**Discussion by udaypalagati** [bpftrace/bpftrace#2092] [Q&A] [answered]
I have found an excellent [article](https://www.gcardone.net/2020-07-31-per-process-bandwidth-monitoring-on-Linux-with-bpftrace/) to track network usage per process using bpftrace, I am convinced to use bpftrace in our linux systems to track the system resource usage per process level but haven't found much details around tracking/reporting CPU and memory per process level. Any inputs on tracking cpu and memory usage  per process using bpftrace would help a lot.

I found following probes around cpu usage, not sure if i can make use of these.

Probes:
kprobe:__cpuusage_read
kprobe:cpuusage_read
kprobe:cpuusage_sys_read
kprobe:cpuusage_user_read
kprobe:cpuusage_write

**Comment by udaypalagati** [comment, marked as answer]
Would also want to know if this is the right candidate for my use case to track the system resource usage(CPU, Memory, Network) on process level

**Reply by danobi** [reply]
Doesn't `top`, `htop`,  or even [below](https://github.com/facebookincubator/below/) do a good job of that? Why bpftrace for cpu and memory?

**Reply by brendangregg** [reply]
Right; There's also sar, and many other metric collectors already (collectd, Prometheus, snap, etc.) that do metric basics like CPU and networking. bpftrace is better used to solve things where there aren't already other tools.

**Reply by udaypalagati** [reply]
Agreed, it will be an overkill to use bpftrace here when tons of other tools are giving what is wanted.
